defi/src/api2/cron-task/dimensions.ts

```diff
@@ -63,14 +63,402 @@ const timeData = {
+  }
+  return { monthly, quarterly }
+}
+
+function rollupFromLabelChart(series: Array<[number, Record<string, number>]>): MQByLabel {
+  const monthly: Record<string, Record<string, number>> = {}
+  const quarterly: Record<string, Record<string, number>> = {}
+  for (const [ts, labels] of series || []) {
+    if (typeof ts !== 'number' || typeof labels !== 'object') continue
+    const mk = monthKeyFromUnix(ts)
+    const qk = quarterKeyFromUnix(ts)
+    for (const [label, v] of Object.entries(labels || {})) {
+      const val = typeof v === 'number' ? v : Number(v) || 0
+      if (!monthly[label]) monthly[label] = {}
+      if (!quarterly[label]) quarterly[label] = {}
+      monthly[label][mk] = (monthly[label][mk] ?? 0) + val
+      quarterly[label][qk] = (quarterly[label][qk] ?? 0) + val
+    }
+  }
+  return { monthly, quarterly }
+}
+
+function readJSONSafe(fp: string): any | null {
+  try { return JSON.parse(fs.readFileSync(fp, 'utf-8')) } catch { return null }
+}
+
+function slugFromAllFile(filename: string) {
+  return filename.endsWith('all') ? filename.slice(0, -3) : filename
+}
+
+function slugFromBlFile(filename: string) {
+  if (filename.endsWith('-bl')) return filename.slice(0, -3)
+  if (filename.endsWith('bl')) return filename.slice(0, -2)
+  return filename
+}
+
+function findBuildDir(): string | null {
+  const candidates = [
+    path.resolve(process.cwd(), '.api2-cache/build'),
+    path.resolve(process.cwd(), 'src/api2/.api2-cache/build'),
+    path.resolve(__dirname, '../../.api2-cache/build'),
+    path.resolve(__dirname, '../../../.api2-cache/build'),
+    path.resolve(__dirname, '../../../../.api2-cache/build'),
+  ]
+  for (const p of candidates) if (fs.existsSync(p)) return p
+  return null
+}
+
+function applyProtocolMetaFromProtoFile(target: any, src: any) {
+  if (!src) return
+  target.metadata  = target.metadata  || {}
+
+  const KEYS = [
+    'id',
+    'defillamaId',
+    'name',
+    'displayName',
+    'slug',
+    'symbol',
+    'url',
+    'logo',
+    'module',
+    'category',
+    'chains',
+    'methodologyURL',
+    'methodology',
+    'gecko_id',
+    'forkedFrom',
+    'twitter',
+    'audits',
+    'audit_links',
+    'description',
+    'address',
+    'cmcId',
+    'github',
+    'governanceID',
+    'treasury',
+    'parentProtocol',
+    'previousNames',
+    'hallmarks',
+    'defaultChartView',
+    'protocolType',
+    'linkedProtocols',
+    'versionKey'
+  ]
+
+  for (const k of KEYS) {
+    if (src[k] !== undefined && target.metadata[k] === undefined) target.metadata[k] = src[k]
+  }
+
+  if (!target.slug && (target.metadata.slug || src.slug)) target.slug = (target.metadata.slug || src.slug)
+}
+
+/*
+   - dssr = df - dr, if dssr missing and df & dr available
+   - dr   = df - dssr, if dr   missing and df & dssr available
+   - dhr  = dr - dssr, if dhr  missing and dr & dssr available
+*/
+function fillDerivedMetrics(proto: AggregatedProtocolOut) {
```

Comment:
imo, this refilling should be done when we are pulling data from db and before it gets added to cache, I would add a new field to protocol cache, `hasBreakdownData`, so if that flag is true or any of the new items have `bl` field, run this logic to auto add missing fields